File: src/CHTL/CHTLParser/Parser.cpp
#include "Parser.h"
#include <sstream>

namespace CHTL {

Parser::Parser(Lexer& lexer) : m_lexer(lexer) {
    // Initialize with two tokens to populate curToken and peekToken
    nextToken();
    nextToken();
}

void Parser::nextToken() {
    m_curToken = m_peekToken;
    m_peekToken = m_lexer.nextToken();
}

bool Parser::currentTokenIs(TokenType type) const {
    return m_curToken.type == type;
}

bool Parser::peekTokenIs(TokenType type) const {
    return m_peekToken.type == type;
}

bool Parser::expectPeek(TokenType type) {
    if (peekTokenIs(type)) {
        nextToken();
        return true;
    } else {
        std::stringstream ss;
        ss << "Error: Expected next token to be " << TokenTypeToString(type)
           << ", but got " << TokenTypeToString(m_peekToken.type) << " instead.";
        m_errors.push_back(ss.str());
        return false;
    }
}

std::unique_ptr<ProgramNode> Parser::parseProgram() {
    auto program = std::make_unique<ProgramNode>();
    while (!currentTokenIs(TokenType::END_OF_FILE)) {
        auto stmt = parseStatement();
        if (stmt) {
            program->children.push_back(stmt);
        }
        nextToken();
    }
    return program;
}

std::shared_ptr<Node> Parser::parseStatement() {
    if (currentTokenIs(TokenType::IDENTIFIER)) {
        // Could be an element or a top-level text/style block.
        // For now, we only support elements at the top level.
        if (m_curToken.literal == "text") {
            return parseTextNode();
        }
        return parseElement();
    } else if (currentTokenIs(TokenType::COMMENT_GEN)) {
        return parseCommentNode();
    }
    return nullptr;
}

std::shared_ptr<ElementNode> Parser::parseElement() {
    auto element = std::make_shared<ElementNode>();
    element->tagName = m_curToken.literal;

    if (!expectPeek(TokenType::LBRACE)) {
        return nullptr;
    }

    nextToken(); // Consume '{'

    while (!currentTokenIs(TokenType::RBRACE) && !currentTokenIs(TokenType::END_OF_FILE)) {
        if (currentTokenIs(TokenType::IDENTIFIER)) {
            if (peekTokenIs(TokenType::COLON) || peekTokenIs(TokenType::EQUALS)) {
                auto attr = parseAttribute();
                if (attr) {
                    element->attributes.push_back(attr);
                }
            } else if (m_curToken.literal == "text") {
                 auto text = parseTextNode();
                 if (text) {
                     element->children.push_back(text);
                 }
            } else {
                auto child = parseElement();
                if (child) {
                    element->children.push_back(child);
                }
            }
        } else if (currentTokenIs(TokenType::COMMENT_GEN)) {
            auto comment = parseCommentNode();
            if (comment) {
                element->children.push_back(comment);
            }
        }
        nextToken();
    }

    if (!currentTokenIs(TokenType::RBRACE)) {
        m_errors.push_back("Error: Unterminated element block, expected '}'.");
        return nullptr;
    }

    return element;
}

std::shared_ptr<AttributeNode> Parser::parseAttribute() {
    auto attribute = std::make_shared<AttributeNode>();
    attribute->name = m_curToken.literal;

    nextToken(); // consume identifier
    if (!currentTokenIs(TokenType::COLON) && !currentTokenIs(TokenType::EQUALS)) {
        m_errors.push_back("Error: Expected ':' or '=' after attribute name.");
        return nullptr;
    }

    nextToken(); // consume ':' or '='

    attribute->value = m_curToken.literal;

    // An unquoted literal might be tokenized as multiple identifiers.
    // This is a simplification for now. The parser will need to be smarter.
    if (!peekTokenIs(TokenType::SEMICOLON)) {
        // Could handle multi-word unquoted literals here if needed in the future.
    } else {
       nextToken(); // consume semicolon
    }


    return attribute;
}

std::shared_ptr<TextNode> Parser::parseTextNode() {
    auto textNode = std::make_shared<TextNode>();

    if (!expectPeek(TokenType::LBRACE)) {
        return nullptr;
    }
    nextToken(); // Consume '{'

    std::stringstream content;
    while (!currentTokenIs(TokenType::RBRACE) && !currentTokenIs(TokenType::END_OF_FILE)) {
        content << m_curToken.literal;
        if (!peekTokenIs(TokenType::RBRACE)) {
            content << " "; // Add space between joined tokens
        }
        nextToken();
    }

    textNode->content = content.str();

    if (!currentTokenIs(TokenType::RBRACE)) {
        m_errors.push_back("Error: Unterminated text block, expected '}'.");
        return nullptr;
    }

    return textNode;
}

std::shared_ptr<CommentNode> Parser::parseCommentNode() {
    auto commentNode = std::make_shared<CommentNode>();
    commentNode->text = m_curToken.literal;
    return commentNode;
}

} // namespace CHTL

File: parser.py
from lexer import Lexer, Token
from nodes import CommentNode, ElementNode, TextNode

class Parser:
    def __init__(self, lexer):
        self.lexer = lexer
        self.tokens = lexer.tokenize()
        self.token_buffer = []

    def peek(self, k=0):
        while len(self.token_buffer) <= k:
            self.token_buffer.append(next(self.tokens))
        return self.token_buffer[k]

    def consume(self, expected_kind=None):
        if not self.token_buffer: self.peek()
        if expected_kind and self.peek().type != expected_kind:
            raise RuntimeError(f"Unexpected token {self.peek()}, expected {expected_kind}")
        return self.token_buffer.pop(0)

    def parse(self):
        root = ElementNode('_root_')
        while self.peek().type != 'EOF':
            root.children.append(self.parse_statement())
        return root

    def parse_statement(self):
        if self.peek().type == 'GEN_CMNT':
            return CommentNode(self.consume().value)
        if self.peek().type == 'ID':
            if self.peek().value == 'text':
                return self.parse_text_element()
            return self.parse_element()
        raise RuntimeError(f"Invalid statement starting with {self.peek()}")

    def parse_element(self):
        tag_name = self.consume('ID').value
        node = ElementNode(tag_name)
        if self.peek().type == 'LBRACE':
            self.consume()
            node.attributes = self.parse_attributes()
            while self.peek().type != 'RBRACE':
                node.children.append(self.parse_statement())
            self.consume('RBRACE')
        return node

    def parse_attributes(self):
        attrs = {}
        while self.peek().type == 'ID' and self.peek(1).type == 'COLON':
            key = self.consume('ID').value
            self.consume('COLON')

            value = ""
            # Consume all tokens until the semicolon
            while self.peek().type != 'SEMI':
                token = self.consume()
                # For quoted strings, use the value directly. For others, add a space.
                if token.type == 'STRING':
                    value = token.value
                else:
                    value += token.value + " "

            attrs[key] = value.strip()
            self.consume('SEMI')
        return attrs

    def parse_text_element(self):
        self.consume('ID') # text
        self.consume('LBRACE')

        value = ""
        # Consume all tokens until the closing brace
        while self.peek().type != 'RBRACE':
            token = self.consume()
            if token.type == 'STRING':
                value = token.value
            else:
                value += token.value + " "

        self.consume('RBRACE')
        return TextNode(value.strip())
